### PR TITLE
[#5619] feat(api): supports credential api for Gravitino

### PR DESCRIPTION
**Pull request by FANNG1**
### What changes were proposed in this pull request?

This is the first PR of credential vending for Gravitino server to add credential API , please refer to #5682 for the overall workflow. Examples to use the API:

```java
    Catalog catalog = metalake.loadCatalog(catalogName);
    Credential[] credentials =
        catalog.supportsCredentials().getCredentials();

    Credential credential =
        catalog.supportsCredentials().getCredential(“s3-token”);
```

```java
    Fileset fileset = catalog.asFilesetCatalog().loadFileset(filesetIdent);
    Credential[] credentials =
        fileset.supportsCredentials().getCredentials();
    Credential credential =
       fileset.supportsCredentials().getCredential(“s3-token”);
 
```

### Why are the changes needed?

Fix: #5619 

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
using draft PR to do E2E test

### CODE REVIEW

**Comment by FANNG1** [comment]
@jerryshao @yuqi1129 please help to review this, thanks

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import java.util.Optional;
+import org.apache.gravitino.exceptions.CredentialDeserializeException;
+import org.apache.gravitino.exceptions.NoSuchCredentialException;
+
+/** Interface to get credentials. */
+public interface SupportsCredentials {
+  /**
+   * Retrieves an array of {@link Credential} objects based on the specified credential type.
+   *
+   * @param credentialType The server will return credentials according to the configuration if
```
Comment (yuqi1129) [inline api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java:31]: You'd better give some example value of `SupportsCredentials`, from your description, I could not infer the possible value of it.

File: api/src/main/java/org/apache/gravitino/exceptions/CredentialDeserializeException.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ *  specific language governing permissions and limitations
+ *  under the License.
+ */
+
+package org.apache.gravitino.exceptions;
+
+import com.google.errorprone.annotations.FormatMethod;
+import com.google.errorprone.annotations.FormatString;
+
+/** An exception thrown when creating the specific credential from credential information failed. */
```
Comment (yuqi1129) [inline api/src/main/java/org/apache/gravitino/exceptions/CredentialDeserializeException.java:25]: The comments is not aligned with the class name.

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+   *     you request a credential for a catalog but the credential provider couldn't generate the
+   *     credential for the catalog, like S3 token credential provider only generate credential for
+   *     the specific object like {@link org.apache.gravitino.file.Fileset}, {@link
+   *     org.apache.gravitino.rel.Table}.
+   * @throws NoSuchCredentialException If the specified credential type cannot be found in the
+   *     server side.
+   * @throws CredentialDeserializeException If there are problems when deserializing the credential
+   *     in client side.
+   */
+  Credential[] getCredentials(Optional<String> credentialType)
```
Comment (yuqi1129) [inline api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java:47]: So we only need parameter `credentialType` to abtain credentails?

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  /**
+   * Retrieves an array of {@link Credential} objects based on the specified credential type.
+   *
+   * @param credentialType The server will return credentials according to the configuration if
+   *     credential type is empty, or else will return the credential with the specified credential
+   *     type.
+   * @return An array of {@link Credential} objects. There will be only one credential for one
+   *     credential type. In most cases the array only contains one credential. If the object like
+   *     {@link org.apache.gravitino.file.Fileset} contains multi locations for different storages
+   *     like HDFS, S3, etc. The array will contain multi credentials. The array could be empty if
```
Comment (yuqi1129) [inline api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java:37]: 1. There will be only one credential for one credential type.
2. In most cases the array only contains one credential
3. If the object like {@link org.apache.gravitino.file.Fileset} contains multi locations for different storages

These three sentences are in conflict with each other.

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java (hunk 3 above)
Comment (FANNG1) [reply api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java:47]: yes, is it necessary to add a more complicated parameter like `CredentialRequest` which contains `credentialType`

File: api/src/main/java/org/apache/gravitino/exceptions/CredentialDeserializeException.java (hunk 2 above)
Comment (FANNG1) [reply api/src/main/java/org/apache/gravitino/exceptions/CredentialDeserializeException.java:25]: `CreateCredentialFailedException`?

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java (hunk 4 above)
Comment (FANNG1) [reply api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java]: updated

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java (hunk 3 above)
Comment (yuqi1129) [reply api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java:49]: I see.

File: api/src/main/java/org/apache/gravitino/exceptions/CredentialDeserializeException.java (hunk 2 above)
Comment (FANNG1) [reply api/src/main/java/org/apache/gravitino/exceptions/CredentialDeserializeException.java]: removed it as the user couldn't do something after catch this exception.

**Comment by FANNG1** [comment]
@jerryshao  could you help to review this, thanks

**Comment by yuqi1129** [comment]
I'm fine with the current implementation.

**Comment by FANNG1** [comment]
@jerryshao  do you have time to review this PR?

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+   *     contains multi locations for different storages like HDFS, S3, the array will contain multi
+   *     credentials. The array could be empty if you request a credential for a catalog but the
+   *     credential provider couldn't generate the credential for the catalog, like S3 token
+   *     credential provider only generate credential for the specific object like {@link
+   *     org.apache.gravitino.file.Fileset}, {@link org.apache.gravitino.rel.Table}. There will be
+   *     at most one credential for one credential type.
+   * @throws NoSuchCredentialException If the specified credential type cannot be found in the
+   *     server side.
+   */
+  Credential[] getCredentials(Optional<String> credentialType) throws NoSuchCredentialException;
```
Comment (jerryshao) [inline api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java:46]: I would suggest using two interfaces:

```
Credentials[] getCredentials();
Credentials getCredentials(String credentialType)
```

The second interface can have a default implementation and  be implemented based on the first one.

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+/** Interface to get credentials. */
+public interface SupportsCredentials {
+
+  /**
+   * Retrieves an array of {@link Credential} objects.
+   *
+   * @return An array of {@link Credential} objects. In most cases the array only contains one
+   *     credential. If the object like {@link org.apache.gravitino.file.Fileset} contains multi
+   *     locations for different storages like HDFS, S3, the array will contain multi credentials.
```
Comment (jerryshao) [inline api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java:37]: Change from "multi" to "multiple" in all the places.

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+   *     The array could be empty if you request a credential for a catalog but the credential
+   *     provider couldn't generate the credential for the catalog, like S3 token credential
+   *     provider only generate credential for the specific object like {@link
+   *     org.apache.gravitino.file.Fileset}, {@link org.apache.gravitino.rel.Table}. There will be
+   *     at most one credential for one credential type.
+   */
+  Credential[] getCredentials() throws NoSuchCredentialException;
+
+  /**
+   * Retrieves an {@link Credential} object based on the specific credential type.
```
Comment (jerryshao) [inline api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java:47]: "... based on the specified credential type."

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+   *
+   * @param credentialType The type of the credential like s3-token, s3-secret-key which defined in
+   *     the specific credentials.
+   * @return An {@link Credential} object with the specific credential type.
```
Comment (jerryshao) [inline api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java:51]: "... with the specified..."

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java (hunk 6 above)
Comment (FANNG1) [reply api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java]: updated

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java (hunk 7 above)
Comment (FANNG1) [reply api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java]: done

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java (hunk 8 above)
Comment (FANNG1) [reply api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java]: done

File: api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    List<Credential> filteredCredentials =
+        Arrays.stream(credentials)
+            .filter(credential -> credentialType.equals(credential.credentialType()))
+            .collect(Collectors.toList());
+    if (filteredCredentials.isEmpty()) {
+      throw new NoSuchCredentialException(
+          "No credential found for the credential type: %s", credentialType);
+    } else if (credentials.length > 1) {
+      throw new IllegalStateException(
+          "Multi credentials found for the credential type:" + credentialType);
```
Comment (jerryshao) [inline api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java:72]: Also change this multi to multiple.

Comment (FANNG1) [reply api/src/main/java/org/apache/gravitino/credential/SupportsCredentials.java]: updated

**Comment by jerryshao** [comment]
@FANNG1 can you please update the PR description to reflect your updated API?

**Comment by FANNG1** [comment]
> @FANNG1 can you please update the PR description to reflect your updated API?

done

**Review by yuqi1129 — APPROVED** [review]

**Pull request merged**